README
Kafka package version SNAPSHOT
==========================

WARNING

This is a fork of the original Kafka Perl modules. This fork is a work in
progress on implementing the new protocol for Kafka 0.8.


INTRO

The Kafka package is a set of Perl modules which provides a
simple and consistent application programming interface (API) to the
Apache Kafka 0.8, a high-throughput distributed messaging system.
The main focus of the package is to provide classes
and functions that allow you to write Apache Kafka clients
without Apache ZooKeeper service.

The Kafka package was written, tested, and found working on 
recent Linux distributions.

INSTALLATION

To install this package type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

In order to install and use this package you will need Perl version
5.010 or better. Some modules within this package depend on other
packages that are distributed separately from Perl. We recommend that
you have the following packages installed before you install
Kafka:

   Digest::CRC
   Params::Util

Digest::CRC version 0.18 has a bug which prevents it from using fast,
XS code if installed via CPAN.
Please install this module manually using
'perl Makefile.PL && make test install' if installed via CPAN shell.

The Kafka package has the following optional dependencies:

   Test::Deep
   Test::Exception

If the optional modules are missing, some "prereq" tests are skipped.

The installation of the missing dependencies can either be accomplished
through your OS package manager or through CPAN (or downloading the source
for all dependencies and compiling them manually).

TESTING

The following environment variables change the way Kafka Benchmark test behave
("t/99_usage/03_kafka_bench.t"):

   # the IO timeout in secs
   KAFKA_BENCHMARK_TIMEOUT = 0.5

   # minimum length of test messages
   KAFKA_BENCHMARK_LEN_MIN = 200

   # maximum length of the test messages
   KAFKA_BENCHMARK_LEN_MAX = 200

   # number of messages for batch testing
   KAFKA_BENCHMARK_PACKAGE = 5000

   # number of messages for a single test
   KAFKA_BENCHMARK_SINGLE = 5

SIMPLE BENCHMARK

Use tools/benchmark_consumer.pl and tools/benchmark_producer.pl to run a simple
benchmark.
Run one or more consumer instances using tools/benchmark_consumer.pl .
and one or more producer instances using tools/benchmark_producer.pl .

DOCUMENTATION

Every Kafka package module has a man page.
A simple description of each module is included below in
the SEE ALSO section.

SEE ALSO

The basic operation of the Kafka package modules:

   Kafka             - constants and messages used by
                       the Kafka package modules
   Kafka::IO         - object interface to socket communications
                       with the Apache Kafka server
   Kafka::Protocol   - functions to process messages
                       in the Apache Kafka's Wire Format
   Kafka::Message    - object interface to the Kafka message properties
   Kafka::Int64      - functions to work with 64 bit elements
                       of the protocol on 32 bit systems
   Kafka::Mock       - object interface to the TCP mock server for testing
   Kafka::Producer   - object interface to the producer client
   Kafka::Consumer   - object interface to the consumer client

COPYRIGHT AND LICENCE

Copyright (C) 2012-2013 by TrackingSoft LLC.
All rights reserved.

This package is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See perlartistic at
http://dev.perl.org/licenses/artistic.html .

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Kafka package

0.23  Fri Mar  28 2014
    CHANGES:
    - Add support for Snappy compression.

0.22  Fri May  24 2013
    CHANGES:
    - Updated Consumer to use BrokerChannels

0.21  Mon May  13 2013
    CHANGES:
    - Initial support for a producer and multiple brokers
    - Addition of BrokerChannels to manage connections to brokers.

0.20  Mon Feb  25 2013
    CHANGES:
    - Initial prototype support for Kafka 0.8 protocol

0.10  Thu Feb  07 2013
    CHANGES:
    - 03_kafka_bench.t does not depend on previous tests

0.09  Mon Jan  14 2013
    CHANGES:
    - benchmark_consumer.pl and benchmark_producer.pl are moved
      to directory tools
    - Makefile.PL by default does not ask questions interactively

0.08  Fri Dec  21 2012
    CHANGES:
    - Tests reorganized

0.07  Thu Dec  18 2012
    FIXES:
    - Checking OS and Perl version in Makefile.PL
      CPAN RT #82018: Tests hang on Windows. (reported by Alexandr Ciornii)

0.06  Wed Sep  26 2012
    CHANGES:
    - Refinement of calculation operation time in benchmarks
    - Some improvements in tests

0.05  Thu Sep  13 2012
    CHANGES:
    - Mock server tests moved to 'xt' directory

0.04  Mon Sep  10 2012
    CHANGES:
    - Author and release tests moved to 'xt' directory

0.03  Thu Aug  30 2012
    CHANGES:
    - Makefile.PL modified for the use of command line
    - Added control code coverage 'coverage.sh'
    - Added test 't/94_fixme.t'
    - Added test 't/95_critic.t'

    BUGFIXES:
    - Fixed version number to 'use 5.010'

0.02  Wed Jun  13 2012
    CHANGES:
    - Minor changes in documentation

    BUGFIXES:
    - Fixed Kafka::Int64 to make it working on Perl 5.12
    - Disallow using 0 timeouts in Kafka::IO and Kafka::Mock

0.01  Tue Jun  12 2012
    - Original version


Makefile.PL
#!/usr/bin/perl -w

use strict;
use warnings;
use 5.010;

if ( $^O eq 'MSWin32' ) {
    print STDERR "OS unsupported: $^O\n";
    exit( 255 );
}

use File::Basename;
use ExtUtils::MakeMaker qw( prompt );
use Getopt::Long;

my ( $no_full_tests, $full_tests, $answer, $topic );

GetOptions(
    'no_full_tests!'    => \$no_full_tests,
    'full_tests!'       => \$full_tests,
    'topic=s'           => \$topic,
);
$answer = 'no'  if $no_full_tests;
$answer = 'yes' if $full_tests;
$answer //= 'no';

my %config = (
    NAME                        => 'Kafka',
    VERSION_FROM                => 'lib/Kafka.pm',  # finds $VERSION
    MIN_PERL_VERSION            => '5.010',
    PREREQ_PM                   => {            # more to the list $Module::CoreList::version{5.010}
        'Digest::CRC'           => 0.18,
        'Params::Util'          => 1.04,
        'Compress::Snappy'      => 0.23,
        },
    BUILD_REQUIRES              => {            # more to the list $Module::CoreList::version{5.010}
        'Test::Deep'            => 0.109,
        'Test::Exception'       => 0.31,
        },
    ( $] >= 5.005 ?                             ## Add these new keywords supported since 5.005
        (
        ABSTRACT        => 'Apache Kafka low-level synchronous API, which does not use Zookeeper.',
        AUTHOR          => 'TrackingSoft LLC <sgladkov@trackingsoft.com>'
        ) : ()
    ),
    dist        => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean       => { FILES => "Debian_CPANTS.txt Kafka-*.tar.gz" },
    LICENSE     => 'perl',
    NO_MYMETA   => 1,
);

my @patterns = qw( t/*.t t/*/*.t );
$config{test} = { TESTS => join ' ', map { glob } @patterns };

print "=  => To run a full tests requires a running local server kafka (port 9092) <=  =\n" unless $answer;
$answer ||= prompt( 'Do you want to run a full tests of the client?', 'no' );

if ( $answer =~ m/^y/i )
{
    print "I'm going to run a full tests.\n";
    $answer = $topic || prompt( 'Please specify the topic for use with the kafka server:', 'test' );

    my @files = $config{test}->{TESTS} =~ m| (t\/99_usage\/\d\d_kafka.*?\.t)|g;
    foreach my $f_name ( @files )
    {
        if ( open( my $fh, "+<", $f_name ) )
        {
            my $content = "";
            while ( <$fh> )
            {
                if ( /^__DATA__/ )
                {
                    chomp $_;
                    $content .= "$_\n$answer";
                    seek $fh, 0, 0;
                    print $fh $content if truncate $fh, 0;
                    last;
                }
                else
                {
                    $content .= $_;
                }
            }
            close $fh;
        }
    }
}
else
{
    $config{test}->{TESTS} =~ s| t\/99_usage\/\d\d_kafka.*?\.t||g;
    print "Miss the interaction with the kafka server.\n";
}

if( $ENV{AUTHOR_TESTS} ) {
    @patterns = qw( xt/author/*.t xt/author/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

if( $ENV{RELEASE_TESTS} ) {
    @patterns = qw( xt/release/*.t xt/release/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

ExtUtils::MakeMaker::WriteMakefile( %config );
